### ISSUE TITLE
sink/pull: receive-side recompression by overriding `compression` property

### ISSUE THREAD

**Issue by problame** [zrepl/zrepl#197]
> It is also reasonable to expect that on a backup system the recordsize parameter would be higher to achieve a better compression ratio, in those cases the data would have to be recompressed anyway, perhaps also with a different algorithm. This is a very important aspect that should also be considered, although not related to compression in transport.

_Originally posted by @siilike in https://github.com/zrepl/zrepl/issues/182#issuecomment-506992436_

It should already be possible to achieve this by setting the properties on `root_fs`, because we neither send properties (`-p`) nor compressed (`-c`).

However, I'm still unclear whether zfs recv recompresses streams generated with `-c`.

**UPDATE**: See comment below https://github.com/zrepl/zrepl/issues/197#issuecomment-512741710 overriding `recordsize` doesn't make sense for this use case.

**Comment by johnramsden** [comment]
Am I understanding correctly that you're trying to determine if a received data set with a different record size will still be compressed?

In my testing it seems so:

```
zfs create vault/test0 -o recordsize=64K -o compression=gzip -o mountpoint=/mnt/test0
for i in $(seq 100); do base64 /dev/urandom | head -c 10000000 > /mnt/test0/$i.txt; done
```

```
zfs get compressratio vault/test0
NAME         PROPERTY       VALUE  SOURCE
vault/test0  compressratio  1.23x  -
```

```
zfs send -c vault/test0 | zfs recv vault/test1
```

```
zfs get compressratio vault/test1
NAME         PROPERTY       VALUE  SOURCE
vault/test1  compressratio  1.23x  -
```

```
zfs get recordsize vault/test1
NAME         PROPERTY    VALUE    SOURCE
vault/test1  recordsize  128K     default
```

If the compression type is different on the parent dataset to the received dataset, the compression type will be changed though. That's not really related to record size though.

**Comment by problame** [comment]
From `man 8 zfs`:

> recordsize=size
> Specifies a suggested block size for files in the file system.  [...]
> database workloads that access files in fixed-size records.  ZFS automatically tunes block sizes according to
> internal algorithms optimized for typical access patterns.
> [...]
> Changing the file system's recordsize affects only files created afterward; existing files are unaffected.

I'm not sure whether compressed recv re-groups the blocks on the receiver according to recordsize, because, AFAIU, this would require decompressing multiple incoming blocks, grouping them together in one new bigger block with the receiver's record size, then re-compressing the new bigger block and writing it out to disk.

I guess you would need to compare the number of blocks in the object that represents each '$i.txt'.
Not sure if that's possible with `zdb`, though.

**Comment by johnramsden** [comment]
I can get the number of indirect blocks in an object with zdb.

```
zdb -O vault/test0 1.txt

    Object  lvl   iblk   dblk  dsize  dnsize  lsize   %full  type
         2    2   128K    64K  7.77M     512  9.56M  100.00  ZFS plain file
```

```
 zdb -ddddd vault/test0 2
Dataset vault/test0 [ZPL], ID 66004, cr_txg 7279552, 777M, 106 objects, rootbp DVA[0]=<0:b6bbeac000:1000> DVA[1]=<0:6eb877d000:1000> [L0 DMU objset] fletcher4 uncompressed unencrypted LE contiguous unique double size=1000L/1000P birth=7281775L/7281775P fill=106 cksum=f3db698e3:2a971d3d41f4:3f64b0c7a7c84a:429b34f6f603a7ea

    Object  lvl   iblk   dblk  dsize  dnsize  lsize   %full  type
         2    2   128K    64K  7.77M     512  9.56M  100.00  ZFS plain file
                                               176   bonus  System attributes
        dnode flags: USED_BYTES USERUSED_ACCOUNTED USEROBJUSED_ACCOUNTED 
        dnode maxblkid: 152
        path    /1.txt
        uid     0
        gid     0
        atime   Sun Jul 14 11:01:16 2019
        mtime   Sun Jul 14 11:01:16 2019
        ctime   Sun Jul 14 11:01:16 2019
        crtime  Sun Jul 14 11:01:16 2019
        gen     7279569
        mode    100644
        size    10000000
        parent  34
        links   1
        pflags  840800000004
Indirect blocks:
               0 L1  0:5c1a039000:2000 20000L/2000P F=153 B=7279569/7279569
               [...]
```

```
zdb -ddddd vault/test1 2
Dataset vault/test1 [ZPL], ID 65863, cr_txg 7279596, 777M, 106 objects, rootbp DVA[0]=<0:b674151000:1000> DVA[1]=<0:76aa24b000:1000> [L0 DMU objset] fletcher4 uncompressed unencrypted LE contiguous unique double size=1000L/1000P birth=7279605L/7279605P fill=106 cksum=f79477e74:2a7e2d01a576:3e8a38345631e7:4160b7611713e120

    Object  lvl   iblk   dblk  dsize  dnsize  lsize   %full  type
         2    2   128K    64K  7.77M     512  9.56M  100.00  ZFS plain file
                                               176   bonus  System attributes
        dnode flags: USED_BYTES USERUSED_ACCOUNTED USEROBJUSED_ACCOUNTED 
        dnode maxblkid: 152
        path    /1.txt
        uid     0
        gid     0
        atime   Sun Jul 14 11:01:16 2019
        mtime   Sun Jul 14 11:01:16 2019
        ctime   Sun Jul 14 11:01:16 2019
        crtime  Sun Jul 14 11:01:16 2019
        gen     7279569
        mode    100644
        size    10000000
        parent  34
        links   1
        pflags  840800000004
Indirect blocks:
               0 L1  0:7e245c2000:2000 20000L/2000P F=153 B=7279597/7279597
               [...]
```

Both datasets 1.txt have 154 indirect blocks. Not sure if that's what we need...

**Comment by problame** [comment]
Well, that suggests to me that the compressed blocks are receiced as is, i.e. no re-grouping.
Could you try the same *without `-c`* (but with different `recordsize` and `comrpession` on the receiving side)?

**Comment by johnramsden** [comment]
```
zfs create vault/testroot -o recordsize=128K -o compression=lz4
zfs send vault/test0 | zfs recv vault/testroot/test1
```

```
zfs get compressratio,compression,recordsize vault/testroot/test1
NAME                  PROPERTY       VALUE     SOURCE
vault/testroot/test1  compressratio  1.00x     -
vault/testroot/test1  compression    lz4       inherited from vault/testroot
vault/testroot/test1  recordsize     128K      inherited from vault/testroot
```

I wasn't really sure what would be useful to you, so just attached the full output of `zdb -dddd` for `vault/test0`, `vault/test1`, and `vault/testroot/test1`.

They appear to be different, but have the same number of objects. Let me know if you want me to parse through them and look for something in particular.

[vault/test0](https://github.com/zrepl/zrepl/files/3400087/zdb-vault-test0.txt)
[vault/test1](https://github.com/zrepl/zrepl/files/3400088/zdb-vault-test1.txt)
[vault/testroot/test1](https://github.com/zrepl/zrepl/files/3400089/zdb-vault-testroot-test1.txt)

**Comment by problame** [comment]
I discussed this feature request with OpenZFS devs:

> Christian Schwarz [11:22 PM]
> I have a zrepl user who asks for `zfs recv -o recordsize=<higher_than_on_source>`. They think that increasing the `recordsize` on their backup server, especially in combination with compression, enables higher compressratio / lower metadata overhead. (https://github.com/zrepl/zrepl/issues/197)Is there any truth to that? AFAIU, a received object's block-size (=recordsize?) is determined by the drr_object.drr_blksz property, and does not change for the lifetime of the object.
> Is that correct?
> Matt Ahrens [11:34 PM]
> That’s correct.  The recordsize on the target has no impact on what `zfs receive` does.  It would only be relevant to files created after the receive (an unusual use case).
> Christian Schwarz [12:06 AM]
> unusual = not happening if the receive-side is only used for further incremental recvs.
> Has anyone thought about the use case proposed by the user? It makes sense, because dict-based compression algos should generally benefit from larger recordsizes (+ obvious lower metadata overhead).
> I imagine the `recv -o recordsize` argument being passed to `receive_writer` as a hint, and `receive_writer` could accumulate drr_write records as long as they are adjacent.
> Matt Ahrens [2:06 AM]
> Yeah it might be possible in some cases. (Plain file, not raw, etc. )
> And we can be sure it’s a new file.

But ATM, this feature is not implemented, hence receiving with a a new `recordsize` has no effect for the recompression use-case.

I'll reduce this issue to just re-compression on the receive-side, which, according to your and my experiments works (although I observed gzip to not recompress if the source stream had 16K recordsize).

**Comment by problame** [comment]
I think this should be possible with #285

**Comment by InsanePrawn** [comment]
Solved (`zfs recv -o compression=lz4`) by #285.
